kernel/src/process/signal/pause.rs

```diff
@@ -22,16 +22,18 @@ use crate::{
 /// On top of `WaitTimeout`, this `Pause` trait provides the `pause`-family methods,
 /// which are similar to the `wait`-family methods except that the methods also return
 /// when the waiting thread is interrupted by a POSIX signal.
-/// When this happens, the `pause`-family methods return `Err(EINTR)`.
+/// When this happens, the `pause`-family methods return `Err(EINTR)` or `Err(ERESTARTSYS)`.
 pub trait Pause: WaitTimeout {
     /// Pauses until the condition is met or a signal interrupts.
     ///
     /// # Errors
     ///
-    /// This method will return an error with [`EINTR`] if a signal is received before the
-    /// condition is met.
+    /// Before the condition is met, this method will return an error with
+    ///  - [`ERESTARTSYS`] if a signal is received and `timeout` is `None`;
+    ///  - [`EINTR`] if a signal is received and `timeout` is `Some(_)`.
```

Comment:
Thank you for the thorough comments! I'd like to reply one by one to see whether we can reach an agreement on whether the changes make sense.

> First, the semantics of the error code is that the system call should be restarted. But this Pause trait is a primitive that is decoupled from any particular system call implementation. So the interface of a Pause object should not make decisions at the system call level.

You are right about all being said but may be wrong with the assumption: this is not about coupling or decoupling things.

The original idea is that a `pause` is like a syscall that can sleep, from a user's perspective (although `pause`'s user is kernel developer and syscall's user is kernel user). They both wait for some events until timing out and may get interrupted, i.e., they are both interruptable waiting.

An interruptable waiting ought to have a mechanism of restarting (no matter what it is, `pause` or syscall) for its caller. The change is not to modify `pause` to adjust to syscall's stuff but to make `pause`'s own semantics complete. It just happens to work well with this PR.

> Second, the comments here are simply invalid. This pause_until method does not even have a timeout argument.

Sorry that's my bad. According to what I stated before, the basic idea is that `pause` should notify its users (here, returning `ERESTARTSYS`) when it can be restarted with exactly the same parameters (including the timer parameter). The comments are not correct here.

> Third, even for pause methods that do have a timeout argument, I don't see why the error code should depend on the input timeout? The caller is fully aware of whether timeout is Some(_) or None. If a Err(EINTR) is returned, the caller can convert the error code, if the conversion is necessary to comply with Linux's behaviors.

It's `pause`'s users' choice whether to restart a `pause`.

You can see it as a syscall returning `EINTR` to its user. At this time the user can choose whether or not to restart. But you cannot just remove the `EINTR` error code from kernel just because some users decide not to restart, which is not fair.

Some syscalls needs special handling logic because they are basically waiting for a signal or doing similar things. They surely don't need a restart since they are just waiting to be interrupted. This is not about a wrong decision being made then got corrected. 

> > I don't quite like the idea of the pause_timeout API. However, I didn't find an appropriate way to get rid of it.

> I don't know your thought process. If you don't like the current Pause interface, fine. Explain why and justify you modifications to the interface. The proposed changes do not make much sense to me at all.

After all I have stated, I have to clarify that I'm not quite against the idea of whole `Pause` trait but the `pause_timeout` method in it. It's just so different from other methods in `Pause` though seems similar. I tried to implement it using `pause_until_or_timeout_impl` (just like any other methods) but didn't find a way.

Reply:
> The original idea is that a `pause` is like a syscall that can sleep, from a user's perspective (although `pause`'s user is kernel developer and syscall's user is kernel user). They both wait for some events until timing out and may get interrupted, i.e., they are both interruptable waiting.

Ok. I got this analog. But I am afraid that it is misleading. For a system call implementation such as futex, it knows whether or not that the system call itself should be restarted. It can make the right decision about returning `ERESTARTSYS` or another return value.

But `Pause` is a general-purpose primitive. It has many users, including many different system calls, each of which has its own Linux behaviors, including system call restart, to conform with. So there is no way for `Pause` to make the right call about restarting an operation or not.

> An interruptable waiting ought to have a mechanism of restarting (no matter what it is, `pause` or syscall) for its caller. The change is not to modify `pause` to adjust to syscall's stuff but to make `pause`'s own semantics complete.

The current set of error codes of `Pause`, which are `ETIME` and `EINTR`,  are sufficient for the caller to make the decision about whether or not it should restart the operation.

> I'm not quite against the idea of whole `Pause` trait but the `pause_timeout` method in it. It's just so different from other methods in `Pause` though seems similar.

I am open to any ideas about improving the `Pause` trait, including its methods with `timeout` argument. But you need to make concrete arguments about its shortcomings. Describing them as being "so different from" is a vague and weak argument. And this is probably also a topic that is out of the scope of this PR.

Reply:
Sorry indeed for the delay. I've been busy these few days.

TL;DR: I'm OK to go in your way if you insist to enable this feature first. But I still prefer the original version a little more though I can now see that the proposed solutions both have their own pros and cons.

> For a system call implementation like futex, it has the context to decide whether the system call should be restarted.

Surely syscalls are capable of deciding their return values on their own. But why should they if we can have a centralized method with exactly the same semantics? In the current implementation, the `ETIME` and `EINTR` returned by a syscall are also directly derived from the return value of the wrapped `Pause`.

We have to distinguish between two concepts: (1) `Pause` make decisions for syscalls who invoke them (2) `Pause`'s return values are just information they want their callers to know and syscalls (and maybe other `Pause` callers) can decide based on the information. I think we are actually implementing the second.  It looks so much like the first just because (1) `Pause` and syscalls share the same set of error codes and (2) syscalls usually just derive the return value, e.g., `ETIME`.

I guess you may be concerned with the fact that sometimes the syscalls themselves just don't want to return `ERESTARTSYS` even though the `Pause` return so. But it's syscalls' obligation to decide what they want (derive the return value of `Pause` or return some other error codes based on it). Suppose we go in your way: keep `Pause` unmodified and let syscalls decide according to`EINTR` returned by `Pause`, then we will also be faced with the same decision: derive the `EINTR` or return `ERESTARTSYS` if it's needed. It doesn't make things any better.

I checked how Linux handles this just now. It turns out Linux's [`wait_event` family](https://elixir.bootlin.com/linux/v6.12.4/source/include/linux/wait.h#L480), who are also general-purpose and are used just like our `Pause`, choose to return `ERESTARTSYS` when interrupted by a signal. They don't even return `EINTR`, which means they don't distinguish between different kinds of interruptions. But wouldn't it be better if we do?